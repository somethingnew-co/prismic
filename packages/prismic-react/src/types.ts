import React from 'react'
import { Elements, HTMLSerializer } from 'prismic-reactjs'
import { LinkResolver } from '@stnew/prismic-types'

export type SliceMap = Map<string, React.ReactType | Promise<any>>

interface Slices {
  [key: string]: React.ReactType | Promise<any>
}

export interface Resolvers {
  linkResolver: LinkResolver
  hrefResolver?: LinkResolver
}

export interface Provider extends Resolvers {
  slices?: Slices
  htmlSerializer?: Hash
}

export interface Context extends Resolvers {
  sliceMap?: SliceMap
  htmlSerializer?: HTMLSerializer<React.ReactNode>
}

// HTMLSerializer types
export interface PropsObject {
  [prop: string]: any
}
export interface Propagator {
  (element?: any): PropsObject
}

export type Hash = {
  [E in Elements]?: [React.ReactType, (PropsObject | Propagator)?]
}
