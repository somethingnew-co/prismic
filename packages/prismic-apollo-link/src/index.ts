export { default as prismicApolloLink } from './apolloLink'
