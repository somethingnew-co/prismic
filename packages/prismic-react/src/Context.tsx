import { createContext } from 'react'
import { Context } from './types'

const prismicContext: Context = {
  slices: {},
  linkResolver: () => '/',
  hrefResolver: () => '/',
  rootResolver: '',
}

const PrismicContext = createContext(prismicContext)

export default PrismicContext
