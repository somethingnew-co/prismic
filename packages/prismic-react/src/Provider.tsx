import React from 'react'
import PrismicContext from './Context'
import { Provider, Context } from './types'

function PrismicProvider({ slices, linkResolver, hrefResolver, children }: Provider): JSX.Element {

  const sliceMap = new Map(Object.entries(slices))

  const value: Context = {
    sliceMap,
    linkResolver,
    hrefResolver,
  }

  return <PrismicContext.Provider value={value}>{children}</PrismicContext.Provider>
}

export default PrismicProvider
