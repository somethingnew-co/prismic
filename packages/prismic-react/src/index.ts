export { default as PrismicContext } from './Context'
export { default as PrismicProvider } from './Provider'
export { default as usePrismic } from './usePrismic'
export { default as useHtmlSerializer } from './useHtmlSerializer'
export { SliceZone } from './SliceZone'
export { default as htmlSerializerThunk } from './utils/serializerThunk'
export * from '@stnew/prismic'
