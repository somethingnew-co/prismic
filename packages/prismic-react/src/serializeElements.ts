import React from 'react'
import { Elements, HTMLSerializer } from 'prismic-reactjs'
import { HTMLSerializerElementMap } from './types'

function serializeElements(map: HTMLSerializerElementMap): HTMLSerializer<React.ReactNode> {
  return function (type, element, content, children, key): React.ReactNode {

    const node = map[type as Elements]

    if (node) {
      const [component, props] = node
      let propagator = props || {}

      if (typeof props === 'function') {
        propagator = props(element)
      }

      const propsWithKey = { ...propagator, key }

      // <img /> elements don't have children
      if (type === Elements.image) {
        return React.createElement(component, propsWithKey)
      }

      return React.createElement(component, propsWithKey, children)
    }
  }
}

export default serializeElements
