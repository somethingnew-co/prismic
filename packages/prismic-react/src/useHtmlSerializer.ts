import React from 'react'
import { HTMLSerializerElementMap } from './types'
import { HTMLSerializer } from 'prismic-reactjs'
import serializerThunk from './utils/serializerThunk'

function useHtmlSerializer(hash: HTMLSerializerElementMap): HTMLSerializer<React.ReactNode> {
  return serializerThunk(hash)
}

export default useHtmlSerializer
