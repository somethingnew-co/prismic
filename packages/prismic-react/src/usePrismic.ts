import { useContext } from 'react'
import PrismicContext from './Context'
import { Context } from './types'

function usePrismic(): Context {
  const value = useContext(PrismicContext)
  return value
}

export default usePrismic
