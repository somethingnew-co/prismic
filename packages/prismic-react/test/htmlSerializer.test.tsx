import React from 'react'
import { RichText, RichTextSpan, Elements } from 'prismic-reactjs'
import { htmlSerializerThunk } from '../src'
import { render } from '@testing-library/react'
import { linkResolver } from '../../../test/resolvers'
import { HTMLSerializerElementMap, HTMLSerializerPropsFunction } from '../src/types'

const linkProps: HTMLSerializerPropsFunction = element => ({
  link: element.data,
})

const imgProps: HTMLSerializerPropsFunction = element => ({
  src: element.url,
  alt: element.alt || '',
  copyright: element.copyright || '',
})

const elementMap: HTMLSerializerElementMap = {
  [Elements.heading1]: ['h1', { 'aria-hidden': true }],
  [Elements.heading2]: ['h2', { className: 'h2' }],
  [Elements.heading3]: ['h3', { className: 'test-class' }],
  [Elements.paragraph]: ['p'],
  [Elements.list]: ['ul', { className: 'list' }],
  [Elements.listItem]: ['li'],
  [Elements.hyperlink]: ['a', linkProps],
  [Elements.image]: ['img', imgProps],
}

export type RichTextBlock = {
  type: Elements
  text: string
  spans: RichTextSpan[]
  alt?: string
  copyright?: string
  dimensions?: { width: number, height: number }
  url?: string
};

interface TextBlock extends RichTextBlock {
  alt?: string
  copyright?: string
  dimensions?: { width: number, height: number }
  url?: string
}

const testData: TextBlock[] = [
  {
    spans: [],
    text: 'Heading 1',
    type: Elements.heading1,
  },
  {
    spans: [],
    text: 'Heading 2',
    type: Elements.heading2,
  },
  {
    spans: [],
    text: 'Heading 3',
    type: Elements.heading3,
  },
  {
    spans: [],
    text: 'If you add another section over here, it splits into a two column layout! ',
    type: Elements.paragraph,
  },
  {
    spans: [
      { start: 12, end: 21, type: 'strong' },
      { start: 22, end: 31, type: 'strong' },
      {
        data: { link_type: 'Web', url: 'https://somethingnew.co', target: '_blank' },
        end: 58,
        start: 46,
        type: 'hyperlink',
      }],
    text: 'And text is formatted correctly. You can even insert links and bulleted lists!',
    type: Elements.paragraph,
  },
  {
    spans: [],
    text: 'Wow, that\'s neat',
    type: Elements.listItem,
  },
  {
    spans: [],
    text: 'Show me more!',
    type: Elements.listItem,
  },
  {
    text: '',
    spans: [],
    alt: 'Alt Text',
    dimensions: { width: 2048, height: 1024 },
    type: Elements.image,
    url: 'https://example.com/image.png',
  },
]


const htmlSerializer = htmlSerializerThunk(elementMap)

test('HTML Serializer', async () => {
  const { container } = render(
    <RichText render={testData} htmlSerializer={htmlSerializer} linkResolver={linkResolver as () => string} />,
  )

  expect(container).toMatchSnapshot()
})
