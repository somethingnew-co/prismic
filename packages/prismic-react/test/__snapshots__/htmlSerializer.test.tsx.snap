// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML Serializer 1`] = `
<div>
  <h1
    aria-hidden="true"
  >
    Heading 1
  </h1>
  <h2
    class="h2"
  >
    Heading 2
  </h2>
  <h3
    class="test-class"
  >
    Heading 3
  </h3>
  <p>
    If you add another section over here, it splits into a two column layout! 
  </p>
  <p>
    And text is 
    <strong>
      formatted
    </strong>
     
    <strong>
      correctly
    </strong>
    . You can even 
    <a
      link="[object Object]"
    >
      insert links
    </a>
     and bulleted lists!
  </p>
  <ul
    class="list"
  >
    <li>
      Wow, that's neat
    </li>
    <li>
      Show me more!
    </li>
  </ul>
  <img
    alt="Alt Text"
    copyright=""
    src="https://example.com/image.png"
  />
</div>
`;
